cmake/modules/extensions.cmake

```
@@ -5393,6 +5384,96 @@ function(zephyr_linker_symbol)
   )
 endfunction()
 
+# Usage:
+#   zephyr_linker_include_generated(CMAKE|KCONFIG|HEADER <name> [PASS <pass>])
+#
+# Add file that is generated at build-time to be included when running the
+# linker script generator.
+#
+# CMAKE <name>     : includes the given cmake file
+# KCONFIG <name>   : import_kconfig() the given Kconfig file. gives
+#                    @variable@ access to all the CONFIG_FOO settings
+# HEADER <name>    : finds all #define FOO value in name. Plain regex, no
+#                    proper preprocessing.
+# PASS [NOT] <pass>: Rule for which PASSES to include file.
+#                    see zephyr_linker_section PASS
+function(zephyr_linker_include_generated)
+  set(single_args "KCONFIG;HEADER;CMAKE")
+  set(multi_args "PASS")
+  cmake_parse_arguments(INCLUDE "" "${single_args}" "${multi_args}" ${ARGN})
+
+  #check that we have exactly one of CMAKE KCONFIG or HEADER
+  set(files)
+  list(APPEND files ${INCLUDE_CMAKE} ${INCLUDE_KCONFIG} ${INCLUDE_HEADER})
+  list(LENGTH files files_count)
+  if(NOT ${files_count} EQUAL 1)
+    message(FATAL_ERROR "zephyr_linker_include_generated(${ARGV0} ...) must have one of CMAKE, KCONFIG or HEADER")
+  endif()
+
+  if(DEFINED INCLUDE_PASS)
+    zephyr_linker_check_pass_param("${INCLUDE_PASS}")
+  endif()
```

afaict what you want to check is that `NOT` is only allowed as the first entry and not as subsequent values, like this:
`INCLUDE_PASS=NOT;SOME;PASS;VALUES`
but you don't care if there is a `NOT` or not.
So those are fine as well:
`INCLUDE_PASS=SOME;PASS;VALUES`
`INCLUDE_PASS=` (see other comment regarding empty values to an argument)
but those should be caught:
`INCLUDE_PASS=SOME;PASS;NOT;VALUES`
`INCLUDE_PASS=NOT;SOME;PASS;NOT;VALUES`

which you can do much easier like this:
```suggestion
  list(SUBLIST INCLUDE_PASS 1 -1 check)
  if("NOT" IN_LIST check)
    message(FATAL_ERROR "NOT only allowed before first <pass> value, like this: PASS [NOT] <pass>...")
  endif()
```

Note: We don't care if `INCLUDE_PASS` is defined or not, because if `INCLUDE_PASS` is undefined or empty, then `check` will also be an empty list when using `SUBLIST`, and thus not contain `NOT`.
Also we don't care the value of the first element, what we care about is that no element after first element is `NOT`.

---

I kept the function, but simplified it using your suggestions.